### PR TITLE
HtmlEditor - add context menu for table operations

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            const $container = $('<div>').appendTo(this.editorInstance.$element());
+            const menuConfig = this._getMenuConfig();
+
+            return this.editorInstance._createComponent($container, ContextMenu, menuConfig);
+        }
+
+        _getMenuConfig(options) {
+            return {
+                target: this._quillContainer,
+                showEvent: 'dxdbclick',
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]: Is it a typo?

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            this._quillContainer = this.editorInstance._getQuillContainer();
+
+            if(this.enabled) {
+                this._contextMenu = this._createContextMenu();
+                this._attachEvents();
+            }
+        }
+
+        _attachEvents() {
+            eventsEngine.on(this.editorInstance._getContent(), CONTEXT_MENU_EVENT, this._openTableContextMenu.bind(this));
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```suggestion
            eventsEngine.on(this.editorInstance._getContent(), CONTEXT_MENU_EVENT, this._prepareContextMenuHandler());
```

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            };
+        }
+
+        _openTableContextMenu(event) {
+            if(this._isTableTarget(event.target)) {
+                this._setContextMenuPosition(event);
+                this._contextMenu.show();
+                event.preventDefault();
+            }
+
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```suggestion
        _prepareContextMenuHandler() {
            return (event) => {
                if(this._isTableTarget(event.target)) {
                  this._setContextMenuPosition(event);
                  this._contextMenu.show();
                  event.preventDefault();
              }
            }

```

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        _setContextMenuPosition(event) {
+            const startPosition = this._quillContainer.get(0).getBoundingClientRect();
+            this?._contextMenu.option({
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]: Is `?.` operator required here? And in the line above?

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                    my: 'left top',
+                    at: 'left top',
+                    collision: 'fit flip',
+                    offset: { x: event.clientX - startPosition.x, y: event.clientY - startPosition.y }
+                }
+            });
+        }
+
+        _isTableTarget(targetElement) {
+            return targetElement.tagName.toUpperCase() === 'TD' || targetElement.tagName.toUpperCase() === 'TH';
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```js
['TD', 'TH'].indexOf(tagName) !== -1
```

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+let TableContextMenuModule = BaseModule;
+
+if(Quill) {
+    TableContextMenuModule = class TableContextMenuModule extends BaseModule {
+        constructor(quill, options) {
+            super(quill, options);
+            this.enabled = !!options.enabled;
+            this._quillContainer = this.editorInstance._getQuillContainer();
+
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js:23]:
```suggestion

this.addCleanCallback(this.prepareCleanCallback());
```

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            if(option === 'enabled') {
+                this.enabled = value;
+                value ? this._attachEvents(true) : this.clean();
+            }
+        }
+
+        clean() {
+            this._detachEvents();
+        }
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```suggestion
        prepareCleanCallback() {
            return () => { this._detachEvents(); };
        }
```

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+        _isTableTarget(targetElement) {
+            return targetElement.tagName.toUpperCase() === 'TD' || targetElement.tagName.toUpperCase() === 'TH';
+        }
+
+        option(option, value) {
+
+            if(option === 'enabled') {
+                this.enabled = value;
+                value ? this._attachEvents(true) : this.clean();
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```suggestion
                value ? this._attachEvents(true) : this._detachEvents();
```

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            return this.editorInstance._createComponent($container, ContextMenu, menuConfig);
+        }
+
+        _getMenuConfig(options) {
+            return {
+                target: this._quillContainer,
+                showEvent: '',
+                dataSource: [
+                    { text: 'Insert', items: [
+                        { text: 'Insert Row Above', onClick: getTableOperationHandler(this.quill, 'insertRowAbove') },
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```suggestion
                        { text: 'Insert Header Row', onClick: getTableOperationHandler(this.quill, 'insertHeaderRow') },
                        { text: 'Insert Row Above', onClick: getTableOperationHandler(this.quill, 'insertRowAbove') },
```
Note: add\use localization constants for text

File: js/ui/html_editor/modules/tableContextMenu.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            const $container = $('<div>').appendTo(this.editorInstance.$element());
+            const menuConfig = this._getMenuConfig();
+
+            return this.editorInstance._createComponent($container, ContextMenu, menuConfig);
+        }
+
+        _getMenuConfig(options) {
+            return {
+                target: this._quillContainer,
+                showEvent: '',
```
Comment (DokaRus) [inline js/ui/html_editor/modules/tableContextMenu.js]:
```suggestion
                showEvent: null,
```

File: js/ui/html_editor/modules/toolbar.js [hunk 11]
```diff
@@ -182,13 +184,13 @@ if(Quill) {
...
                 subscript: this._prepareShortcutHandler('script', 'sub'),
                 insertTable: this._prepareInsertTableHandler(),
-                insertRowAbove: this._getTableOperationHandler('insertRowAbove'),
-                insertRowBelow: this._getTableOperationHandler('insertRowBelow'),
-                insertColumnLeft: this._getTableOperationHandler('insertColumnLeft'),
-                insertColumnRight: this._getTableOperationHandler('insertColumnRight'),
-                deleteColumn: this._getTableOperationHandler('deleteColumn'),
-                deleteRow: this._getTableOperationHandler('deleteRow'),
-                deleteTable: this._getTableOperationHandler('deleteTable')
+                insertRowAbove: getTableOperationHandler(this.quill, 'insertRowAbove'),
```
Comment (DokaRus) [inline js/ui/html_editor/modules/toolbar.js:188]:
```suggestion
                insertHeaderRow: getTableOperationHandler(this.quill, 'insertHeaderRow'),
                insertRowAbove: getTableOperationHandler(this.quill, 'insertRowAbove'),
```

**Review by AlexJameson — DISMISSED** [review]

**Review by DokaRus — DISMISSED** [review]

**Review by Zedwag — APPROVED** [review]

**Review by DokaRus — APPROVED** [review]

**Pull request merged**